crates/optimism/consensus/src/validation/isthmus.rs

```diff
@@ -0,0 +1,127 @@
+//! Block verification w.r.t. consensus rules new in Isthmus hardfork.
+
+use crate::OpConsensusError;
+use alloy_consensus::BlockHeader;
+use core::fmt;
+use reth_optimism_primitives::predeploys::ADDRESS_L2_TO_L1_MESSAGE_PASSER;
+use reth_storage_api::StorageRootProvider;
+use reth_trie_common::HashedStorage;
+use revm::db::BundleAccount;
+
+/// Verifies that `withdrawals_root` (i.e. `l2tol1-msg-passer` storage root since Isthmus) field is
+/// set in block header.
+pub fn verify_withdrawals_storage_root_is_some<H: BlockHeader + fmt::Debug>(
+    header: H,
+) -> Result<(), OpConsensusError> {
+    header.withdrawals_root().ok_or(OpConsensusError::L2WithdrawalsRootMissing)?;
+
+    Ok(())
+}
+
+/// Verifies block header field `withdrawals_root` against storage root of
+/// `L2ToL1MessagePasser.sol` predeploy post block execution.
+///
+/// See <https://specs.optimism.io/protocol/isthmus/exec-engine.html#l2tol1messagepasser-storage-root-in-header>.
+pub fn verify_withdrawals_storage_root<DB, H>(
+    predeploy_account_updates: Option<&BundleAccount>,
+    state: DB,
+    header: H,
+) -> Result<(), OpConsensusError>
+where
+    DB: StorageRootProvider,
+    H: BlockHeader + fmt::Debug,
+{
+    let header_storage_root =
+        header.withdrawals_root().ok_or(OpConsensusError::L2WithdrawalsRootMissing)?;
+
+    // if block contained l2 withdrawals transactions, use predeploy storage updates from
+    // execution
+    let hashed_storage_updates = predeploy_account_updates.map(|acc| {
+        HashedStorage::from_plain_storage(
+            acc.status,
+            acc.storage.iter().map(|(slot, value)| (slot, &value.present_value)),
+        )
+    });
+
+    let storage_root = state
+        .storage_root(ADDRESS_L2_TO_L1_MESSAGE_PASSER, hashed_storage_updates.unwrap_or_default())
+        .map_err(OpConsensusError::L2WithdrawalsRootCalculationFail)?;
+
+    if header_storage_root != storage_root {
+        return Err(OpConsensusError::L2WithdrawalsRootMismatch {
+            header: header_storage_root,
+            exec_res: storage_root,
+        })
+    }
```

Comment:
@rkrasiuk could you ptal at this

Reply:
or @clabby if @rkrasiuk doesn't have capacity. waiting since last week for roman to have capacity to review this https://github.com/paradigmxyz/reth/pull/14307

Reply:
(tag you're teams evm guy, lol)

Reply:
the API usage makes sense